csrc/scheduler/reduction_tma.cpp

```diff
@@ -6,26 +6,269 @@
  */
 // clang-format on
 
+#include <ATen/cuda/CUDAContext.h>
+#include <scheduler/cache_policy_refiner.h>
 #include <scheduler/reduction_tma.h>
+#include <scheduler/reduction_utils.h>
+#include <scheduler/runtime_info.h>
+#include <scheduler/tools/inlining.h>
+#include <scheduler/utils.h>
 
 namespace nvfuser {
 namespace reduction {
 namespace tma {
 
-std::unique_ptr<ReductionParams> getReductionHeuristics(
+std::unique_ptr<TmaInnerReductionParams> getReductionHeuristics(
     Fusion* fusion,
     SchedulerRuntimeInfo& runtime_info,
-    HeuristicDataCache* data_cache) {
+    HeuristicDataCache* data_cache,
+    const reduction_scheduler_utils::FusionRuntimeProperties& props) {
   FusionGuard fg(fusion);
-  auto params = std::make_unique<ReductionParams>();
+
+  // Get device properties
+  auto dev_prop = at::cuda::getCurrentDeviceProperties();
+  const int64_t max_threads_per_sm = dev_prop->maxThreadsPerMultiProcessor;
+  const int64_t target_threads_per_sm = max_threads_per_sm / 2;
+  const int64_t threads_per_warp = dev_prop->warpSize;
+
+  const int64_t smem_elems = dev_prop->sharedMemPerBlockOptin /
```

Comment:
move smem check to `mayUseTma`, depends on number of inputs, we can assert/check there is only 1 input in this PR.